src/db.c

```diff
@@ -324,6 +324,24 @@ int getKeySlot(sds key) {
     return calculateKeySlot(key);
 }
 
+/* Return the slot of the key in the command. IO threads use this function
+ * to calculate slot to reduce main-thread load */
+int getSlotFromCommand(struct redisCommand *cmd, robj **argv, int argc) {
+    int slot = -1;
+    if (!cmd || !server.cluster_enabled) return slot;
+
+    /* Get the keys from the command */
+    getKeysResult result = GETKEYS_RESULT_INIT;
+    int numkeys = getKeysFromCommand(cmd, argv, argc, &result);
```

Comment:
if we do that, maybe we can cache the result so it can serve others usages (ACL, Cluster, ROF).
besides, maybe instead of just computing the slot of the first key, it can already check for cross slot, and then the main thread won't have to.
we did that in lookahead.

Reply:
yes. we can let IO threads do this and cache the result. This PR is based on https://github.com/valkey-io/valkey/pull/861, valkey hasn’t optimized for this issue, so I kept the original behavior, and i don't want to make it bigger, maybe a new PR is better, as you said, we can cache the key result, even calculate if all keys are in a single slot in the IO thread.

besides, I have tried to cache key result but there is no significant performance improvement, so i think it is not urgent.

Reply:
no significant performance improvement.. did you check with cluster mode or ACL?
seems odd that we invest if offloading argument parsing, and command lookup, but claim that offloading key name extraction and slot number / cross slot isn't significant.

in any case, i don't mind taking it in a separate PR, arguing this one is about memory prefetching only. the reason i commented was because this change does add a call to getKeysFromCommand, so essentially adding some work that's now done twice.

anyway, if / when we'll combine the lookahead project this this, we'll get it cached anyway.

Reply:
> anyway, if / when we'll combine the lookahead project this this, we'll get it cached anyway.

since the lookahead project did this, we can have this feature (cache keys result) after merging it?

Reply:
not sure i understand, 0.1% of the main thread CPU in which scenario? if there's no ACL or cluster, it's not used by the main thread.

in any case, we agree it should be cached, and we agree it can be done later. the only question is if we have some small regression because we can now call it twice (e.g. in cluster mode). i guess we don't care as long as the impact of the PR is still positive and we have a plan to improve later.

Reply:
Yes. Keys result can be used
- memory prefetch
- slot calculating for cluster/cluster-compatibility
- ACL check

And as oran said, the IO thread also should check if all keys are in the same slot, instead of only the first key, so the main thread can just report cross-slot error (MULTI-EXEC requires special treatment), so i want to put these works in a separate PR (Valkey also intends to do this, but hasn’t implemented it yet). lookahead project already did this, maybe we just reuse this logic when merging. Besides, as i showed above,  the regression  of calling  `getKeysFromCommand` is not notable.